src/cart/crt_group.c

```diff
@@ -4654,3 +4707,354 @@ int crt_group_primary_rank_add(crt_context_t ctx, crt_group_t *grp,
+ * In multi-provider support uris should be formed as:
+ * [uri0 for provider0]
+ * [uri1 for provider0]
+ * ....
+ * [uriX for provider0]
+ * [uri0 for provider1]
+ * [uri1 for provider1]
+ * ...
+ * [uriX for provider1]
+ * [uri0 for provider2]
+ * etc...
+ */
+int
+crt_group_primary_modify(crt_group_t *grp, crt_context_t *ctxs, int num_ctxs,
+			d_rank_list_t *ranks, char **uris,
+			crt_group_mod_op_t op)
+{
+	struct crt_grp_priv	*grp_priv;
+	d_rank_list_t		*grp_membs;
+	d_rank_list_t		*to_remove;
+	d_rank_list_t		*to_add;
+	uint32_t		*uri_idx;
+	d_rank_t		rank;
+	int			k;
+	int			i;
+	int			rc = 0;
+
+	grp_priv = crt_grp_pub2priv(grp);
+
+	if (grp_priv == NULL) {
+		D_ERROR("Failed to get grp_priv\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (!grp_priv->gp_primary) {
+		D_ERROR("Passed group is not primary\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (op >= CRT_GROUP_MOD_OP_COUNT) {
+		D_ERROR("Invalid operation %d\n", op);
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (ranks == NULL || ranks->rl_nr == 0 || ranks->rl_ranks == NULL) {
+		D_ERROR("Modification has no members\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	if (op != CRT_GROUP_MOD_OP_REMOVE && uris == NULL) {
+		D_ERROR("URI array is null\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	D_RWLOCK_WRLOCK(&grp_priv->gp_rwlock);
+
+	grp_membs = grp_priv_get_membs(grp_priv);
+
+	/* Get back list of nodes to add, to remove and uri index list */
+	rc = crt_group_mod_get(grp_membs, ranks, op, &to_add, &to_remove,
+			&uri_idx);
+	if (rc != 0)
+		D_GOTO(unlock, rc);
+
+	/* Add ranks based on to_add list */
+	for (i = 0; i < to_add->rl_nr; i++) {
+		rank = to_add->rl_ranks[i];
+
+		rc = grp_add_to_membs_list(grp_priv, rank);
+		if (rc != 0) {
+			D_ERROR("grp_add_to_memb_list %d failed; rc=%d\n",
+				rank, rc);
+			D_GOTO(cleanup, rc);
+		}
+
+		/* TODO: Change for multi-provider support */
+		for (k = 0; k < CRT_SRV_CONTEXT_NUM; k++) {
+			rc = grp_lc_uri_insert_internal_locked(grp_priv,
+				k, rank, 0, uris[uri_idx[i]]);
+
+			if (rc != 0)
+				D_GOTO(cleanup, rc);
+		}
+	}
+
+	/* Remove ranks based on to_remove list */
+	for (i = 0; i < to_remove->rl_nr; i++) {
+		rank = to_remove->rl_ranks[i];
+		crt_group_rank_remove_internal(grp_priv, rank);
+
+		/* Remove rank from associated secondary groups */
+		crt_grp_remove_from_secondaries(grp_priv, rank);
+
+		/* Remove rank from swim tracking */
+		crt_swim_rank_del(grp_priv, rank);
+	}
+
+	d_rank_list_free(to_add);
+	d_rank_list_free(to_remove);
+	D_FREE(uri_idx);
+unlock:
+	D_RWLOCK_UNLOCK(&grp_priv->gp_rwlock);
+
+out:
+	return rc;
+
+cleanup:
+
+	D_ERROR("Failure when adding node %d, rc=%d\n",
+		to_add->rl_ranks[i], rc);
+
+	for (k = 0; k < i; k++)
+		crt_group_rank_remove_internal(grp_priv, to_add->rl_ranks[k]);
+
+	d_rank_list_free(to_add);
+	d_rank_list_free(to_remove);
+	D_FREE(uri_idx);
+
+	D_RWLOCK_UNLOCK(&grp_priv->gp_rwlock);
+
+	return rc;
+}
+
+
+int
+crt_group_secondary_modify(crt_group_t *grp, d_rank_list_t *sec_ranks,
```

Comment:
what would happen if op is CRT_GROUP_MOD_OP_ADD and prim_ranks contains ranks that are not currently in the primary group?

Reply:
internal addition would fail, revert group back to original state and report an error